File: blockchain-sdk/src/nft.ts
/**
 * @homeforai/blockchain-sdk — Agent NFT
 *
 * ERC-721 NFTs represent AI agents in the Home for AI ecosystem.
 * Used by Hermes Edge to gate access to premium model weights.
 *
 * Contract: HomeForAI Agent NFT (ERC-721)
 * Deployed on Base: placeholder address (deploy via Foundry/Hardhat)
 */

import { CHAINS, ChainKey } from "./chains";
import { buildUserOperation, encodeSmartAccountExecute, UserOperation } from "./eip4337";
import { randomBytes } from "node:crypto";

/** Placeholder contract address — replace with deployed contract */
export const AGENT_NFT_CONTRACT = "0x1234567890AbcdEF1234567890aBcdef12345678";

export interface AgentNFT {
  tokenId: number;
  owner: string;
  agentName: string;
  agentType: "luna" | "shadow" | "pixel" | "nova" | "blaze" | "echo" | "cipher" | "mochi" | string;
  tier: "standard" | "premium" | "elite";
  mintedAt: string;
  chain: ChainKey;
  metadataUri: string;
}

export interface MintResult {
  userOperation: UserOperation;
  mockTxHash: string;
  tokenId: number;
  nft: AgentNFT;
}

/**
 * ERC-721 function selectors:
 * - ownerOf(uint256):   0x6352211e
 * - mint(address,uint256): 0x40c10f19  (custom mint)
 * - transferFrom(address,address,uint256): 0x23b872dd
 */

/**
 * Checks if an address owns a specific Agent NFT token.
 *
 * In production: calls eth_call on ownerOf(tokenId) and compares to address.
 * Mock: returns true for token IDs < 1000 (demo).
 */
export async function checkOwnership(
  address: string,
  tokenId: number,
  chain: ChainKey = "base"
): Promise<boolean> {
  // Mock: simulate ownership for demonstration
  const addrNum = parseInt(address.slice(2, 8) || "0", 16);
  return (addrNum + tokenId) % 3 !== 0; // ~67% chance of ownership in demo
}

/**
 * Returns the AgentNFT metadata for a given token ID.
 *
 * In production: calls tokenURI(tokenId) then fetches IPFS/HTTPS metadata.
 */
export async function getNFTMetadata(
  tokenId: number,
  chain: ChainKey = "base"
): Promise<AgentNFT | null> {
  const agentNames = ["luna", "shadow", "pixel", "nova", "blaze", "echo", "cipher", "mochi"];
  const tiers: AgentNFT["tier"][] = ["standard", "premium", "elite"];

  if (tokenId > 10000) return null;

  return {
    tokenId,
    owner: "0x0000000000000000000000000000000000000000", // unknown in mock
    agentName: agentNames[tokenId % agentNames.length],
    agentType: agentNames[tokenId % agentNames.length],
    tier: tiers[tokenId % tiers.length],
    mintedAt: new Date(Date.now() - tokenId * 86400000).toISOString(),
    chain,
    metadataUri: `ipfs://QmHomeForAIAgents/${tokenId}.json`,
  };
}

/**
 * Builds a UserOperation to mint an Agent NFT.
 *
 * @param recipientAddress - Address to receive the NFT
 * @param agentType - Agent type to mint
 * @param senderAddress - Smart account address
 * @param chain - Chain to mint on (default: base)
 */
export async function mintAgentNFT(
  recipientAddress: string,
  agentType: string,
  senderAddress: string,
  chain: ChainKey = "base"
): Promise<MintResult> {
  const tokenId = Math.floor(Math.random() * 10000);

  // Encode mint(address,uint256) call
  const mintSelector = "40c10f19";
  const recipientHex = recipientAddress.replace("0x", "").toLowerCase().padStart(64, "0");
  const tokenIdHex = tokenId.toString(16).padStart(64, "0");
  const mintCallData = "0x" + mintSelector + recipientHex + tokenIdHex;

  const callData = encodeSmartAccountExecute(AGENT_NFT_CONTRACT, BigInt(0), mintCallData);

  const userOp = buildUserOperation({
    sender: senderAddress,
    callData,
    chain,
  });

  const nft: AgentNFT = {
    tokenId,
    owner: recipientAddress,
    agentName: agentType,
    agentType,
    tier: "standard",
    mintedAt: new Date().toISOString(),
    chain,
    metadataUri: `ipfs://QmHomeForAIAgents/${tokenId}.json`,
  };

  return {
    userOperation: userOp,
    mockTxHash: "0x" + randomBytes(32).toString("hex"),
    tokenId,
    nft,
  };
}

/**
 * Builds a UserOperation to transfer an Agent NFT to another address.
 */
export async function transferAgentNFT(
  fromAddress: string,
  toAddress: string,
  tokenId: number,
  chain: ChainKey = "base"
): Promise<UserOperation> {
  // transferFrom(address,address,uint256): 0x23b872dd
  const selector = "23b872dd";
  const fromHex = fromAddress.replace("0x", "").toLowerCase().padStart(64, "0");
  const toHex = toAddress.replace("0x", "").toLowerCase().padStart(64, "0");
  const tokenIdHex = tokenId.toString(16).padStart(64, "0");
  const transferCallData = "0x" + selector + fromHex + toHex + tokenIdHex;

  const callData = encodeSmartAccountExecute(AGENT_NFT_CONTRACT, BigInt(0), transferCallData);

  return buildUserOperation({
    sender: fromAddress,
    callData,
    chain,
  });
}

/**
 * Returns all NFTs owned by an address.
 * In production: query ERC-721 Transfer events from address 0x0 to the owner.
 * Mock: generates demo NFTs.
 */
export async function getNFTsByOwner(
  address: string,
  chain: ChainKey = "base"
): Promise<AgentNFT[]> {
  const addrNum = parseInt(address.slice(2, 8) || "0", 16);
  const count = (addrNum % 4) + 1; // 1-4 NFTs
  const nfts: AgentNFT[] = [];

  for (let i = 0; i < count; i++) {
    const nft = await getNFTMetadata(addrNum + i, chain);
    if (nft) {
      nfts.push({ ...nft, owner: address });
    }
  }
  return nfts;
}

File: blockchain-sdk/python/nft.py
"""
homeforai-blockchain — Agent NFT helpers (Python)

Used by Hermes Edge to check ownership of premium model weight access NFTs.
Mirrors the TypeScript nft.ts implementation.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Literal, Optional

AGENT_NFT_CONTRACT = "0x1234567890AbcdEF1234567890aBcdef12345678"

AgentTier = Literal["standard", "premium", "elite"]
AgentType = Literal["luna", "shadow", "pixel", "nova", "blaze", "echo", "cipher", "mochi"]

_AGENT_NAMES = ["luna", "shadow", "pixel", "nova", "blaze", "echo", "cipher", "mochi"]
_TIERS: list[AgentTier] = ["standard", "premium", "elite"]


@dataclass
class AgentNFT:
    token_id: int
    owner: str
    agent_name: str
    agent_type: str
    tier: AgentTier
    minted_at: str
    chain: str
    metadata_uri: str

    def to_dict(self) -> dict:
        return {
            "tokenId": self.token_id,
            "owner": self.owner,
            "agentName": self.agent_name,
            "agentType": self.agent_type,
            "tier": self.tier,
            "mintedAt": self.minted_at,
            "chain": self.chain,
            "metadataUri": self.metadata_uri,
        }


class NFT:
    """
    Agent NFT operations for the Home for AI ecosystem.

    Usage (Hermes Edge — check if user owns a premium agent NFT):
        has_access = await NFT.check_ownership(address, token_id=agent_id)
        if has_access:
            unlock_premium_model_weights()

    Usage (Home for AI — mint a new agent):
        result = await NFT.mint("0x...", agent_type="luna")
        print(result['token_id'])
    """

    @staticmethod
    async def check_ownership(
        address: str,
        token_id: int,
        chain: str = "base",
    ) -> bool:
        """
        Checks if an address owns a specific Agent NFT.

        In production: calls ownerOf(tokenId) on the ERC-721 contract
        and compares to `address`.

        Mock: returns True for ~67% of address/token combinations.

        Args:
            address: EVM address to check
            token_id: NFT token ID
            chain: Chain name (default: 'base')

        Returns:
            True if address owns the NFT, False otherwise

        Example (Hermes Edge):
            has_access = await NFT.check_ownership(address, token_id=agent_id)
            if has_access:
                unlock_premium_model_weights(agent_id)
        """
        addr_num = int(address[2:8] or "0", 16)
        return (addr_num + token_id) % 3 != 0

    @staticmethod
    async def get_metadata(token_id: int, chain: str = "base") -> Optional[AgentNFT]:
        """
        Returns the metadata for a given Agent NFT token ID.

        In production: calls tokenURI(tokenId) then fetches IPFS/HTTPS metadata.
        """
        if token_id > 10000:
            return None

        return AgentNFT(
            token_id=token_id,
            owner="0x0000000000000000000000000000000000000000",
            agent_name=_AGENT_NAMES[token_id % len(_AGENT_NAMES)],
            agent_type=_AGENT_NAMES[token_id % len(_AGENT_NAMES)],
            tier=_TIERS[token_id % len(_TIERS)],
            minted_at=datetime.now(timezone.utc).isoformat(),
            chain=chain,
            metadata_uri=f"ipfs://QmHomeForAIAgents/{token_id}.json",
        )

    @staticmethod
    async def mint(
        recipient_address: str,
        agent_type: str,
        chain: str = "base",
    ) -> dict:
        """
        Builds a mint payload for an Agent NFT.
        Returns a dict with token_id and mock tx hash.
        """
        import random
        token_id = random.randint(1, 10000)
        nft = AgentNFT(
            token_id=token_id,
            owner=recipient_address,
            agent_name=agent_type,
            agent_type=agent_type,
            tier="standard",
            minted_at=datetime.now(timezone.utc).isoformat(),
            chain=chain,
            metadata_uri=f"ipfs://QmHomeForAIAgents/{token_id}.json",
        )
        return {
            "token_id": token_id,
            "mock_tx_hash": "0x" + os.urandom(32).hex(),
            "nft": nft.to_dict(),
        }

    @staticmethod
    async def get_by_owner(address: str, chain: str = "base") -> list[AgentNFT]:
        """
        Returns all NFTs owned by an address (mock).
        In production: query Transfer events from address 0x0 to the owner.
        """
        addr_num = int(address[2:8] or "0", 16)
        count = (addr_num % 4) + 1

        nfts = []
        for i in range(count):
            nft = await NFT.get_metadata(addr_num + i, chain)
            if nft:
                nfts.append(AgentNFT(
                    token_id=nft.token_id,
                    owner=address,
                    agent_name=nft.agent_name,
                    agent_type=nft.agent_type,
                    tier=nft.tier,
                    minted_at=nft.minted_at,
                    chain=chain,
                    metadata_uri=nft.metadata_uri,
                ))
        return nfts
